src/sync/lookupApiUtils.js

```diff
@@ -32,6 +37,18 @@ const PARAMETERS = {
   registrationCode: { key: 'code', type: TYPES.STRING },
 };
 
+const bugsnagClient = new BugsnagClient();
+
+class BugsnagError extends Error {
+  constructor(message, data, ...args) {
+    super(message, ...args);
+    bugsnagClient.notify(this, report => {
+      report.errorMessage = message;
+      report.metadata = data;
+    });
+  }
+}
```

Comment:
Nice. Do you think `notify` shouldn't be in the constructor? What's the point of this being a class?

Reply:
Just for consistency with other "native" errors: see "Custom Error Types" [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error).

Reply:
Ah yup. still not sure if the notify needs to be in the constructor rather than 

`const newBugsnag = new BugsnagError(error)`
`newBugsnag.notify()`

Two actions so making it explicit in two calls - personal preference

Reply:
I just realised it would be more like

`const newBugsnag = new BugsnagError(error)`
`const newBugsang.notify()`
`throw newBugsnag`

Basically, I am questioning the need for this to be an error type rather than a creating a `notify` wrapper:

`notifyWrapper(error)`
`throw error`

but again I think personal preference

Reply:
Hm. I'm not seeing the raising of an error in the constructor of "Custom Error Types" and so agree with Josh's interpretation soz.

If the error is notified in the constructor, then I think the naming should reflect that
`BugsnagError` => ctor returns an object
`BugsnagNotification` => ctor could create an object and raise a notification
perhaps.

Reply:
Just succinctly, it's just polymorphism, i.e. an `Error` should behave like an error (so having a `notify` method on an error doesn't seem like a good idea). A wrapper seems a bit overkill, as the basic idea here is to be able to throw an error as you would normally, but also have it log to bugsnag (so if you wanted to have a specific bugsnag error, you could then inherit from that). I think in this case inheritance is better than composition (considering that we are talking about OOP). I would agree it's mainly personal preference, but I think it's pretty consistent with standard OOP practices. 

@mark-prins Sorry, not following! The constructor doesn't raise any errors, the object is an error itself (so when it's raised, it also logs upstream by reporting the error to bugsnag). It's polymorphism, pretty much - it walks like an error, it talks like an error, so it is an error,,, but it also happens to log to bugsnag!

Reply:
perhaps it's my wording. by 'raise an error' in the ctor I mean the call to `bugsnagClient.notify` (i.e. logging upstream). 

when you say 'raise an error' - do you mean create the error object? by which - calling `notify` in the ctor is all good. I was thinking that the ctor should just construct the error object. The `notify()` should be done when the error is `throw`n - that's how I thought the process usually runs.

all of which sounds exactly like your response to @joshxg - so perhaps we're saying the same thing?

Reply:
@mark-prins I agree with that, and I think we are all basically saying the same thing, which is that doing "logic" on instantiation seems wrong, so the notification should be decoupled from the constructor. Ideally, the error should notify when it is thrown, but the only way to really do that is when it is instantiated (as far as I know!). If there was a way to fire the notify on `throw` (dynamically, i.e. not explicitly in the calling code), that would be ideal, but I don't think there is? 

Maybe helps to clarify that the main reason for the custom error is to avoid having to have a conditional on certain errors (before they are thrown or after they are caught). The goal is for the error to be thrown just like any other error, but also to notify in the process (without the calling code having to call any methods that only exist on that subclass).

Does that make sense?

Reply:
yep. all good - I don't know of any way to extend `throw` so this is probably as good as it gets